OpenRA.Mods.Common/Traits/World/ShroudRenderer.cs

```diff
@@ -297,13 +294,19 @@ void IRenderShroud.RenderShroud(Shroud shroud, WorldRenderer wr)
 				fogLayer.Update(uv, fogSprite, fogPos);
 			}
 
-			cellsDirty.Clear();
-			cellsAndNeighborsDirty.Clear();
-
 			fogLayer.Draw(wr.Viewport);
 			shroudLayer.Draw(wr.Viewport);
 		}
 
+		void UpdateShroudCell(MPos uv)
+		{
+			var uvp2 = new MPos(uv.U + 2, uv.V + 2);
+			cellsDirty[uvp2] = true;
+			var cell = uvp2.ToCPos(map);
+			foreach (var direction in CVec.Directions)
+				cellsDirty[cell + direction] = true;
```

Comment:
`Map.Contains(PPos)` can be used to check whether a projected cell is inside the map (and therefore the also the CellLayer).

Reply:
Updated.